File: built-in plugins dsp/mb-comp-bus.cpp
/**
 * MB Bus Comp
 * Category : effect
 * Type     : compressor
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Mix bus glue compressor
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_COMP_BUS_H
#define MB_COMP_BUS_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbCompBus : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-comp-bus";
    static constexpr const char* PLUGIN_NAME    = "MB Bus Comp";
    static constexpr const char* PLUGIN_TYPE    = "compressor";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float threshold = -12f;  // range [-40, 0]
    float ratio = 2f;  // range [1, 8]
    float attack = 30f;  // range [0.1, 100]
    float release = 300f;  // range [50, 1000]
    };

    MbCompBus() = default;
    ~MbCompBus() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.threshold = std::clamp(params.threshold, -40f, 0f);
        params.ratio = std::clamp(params.ratio, 1f, 8f);
        params.attack = std::clamp(params.attack, 0.1f, 100f);
        params.release = std::clamp(params.release, 50f, 1000f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Bus Comp
        return input;
    }
};

#endif // MB_COMP_BUS_H

File: built-in plugins dsp/mb-comp-bus.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbCompBusPlugin: PluginDefinition = { id: 'mb-comp-bus', slug: 'mb-comp-bus', name: 'MB Bus Comp', category: 'effect', type: 'compressor', version: '1.0.0', description: 'Mix bus glue compressor', author: 'Max Booster', parameters: [{ id: 'threshold', name: 'Threshold', type: 'float', defaultValue: -12, minValue: -40, maxValue: 0, automatable: true }, { id: 'ratio', name: 'Ratio', type: 'float', defaultValue: 2, minValue: 1, maxValue: 8, automatable: true }, { id: 'attack', name: 'Attack', type: 'float', defaultValue: 30, minValue: 0.1, maxValue: 100, automatable: true }, { id: 'release', name: 'Release', type: 'float', defaultValue: 300, minValue: 50, maxValue: 1000, automatable: true }], defaultPreset: { threshold: -12, ratio: 2, attack: 30, release: 300 } };

export default MbCompBusPlugin;
